byterun/hash.c

```diff
@@ -273,7 +277,20 @@ CAMLprim value caml_hash(value count, value limit, value seed, value obj)
         /* Mix in the tag and size, but do not count this towards [num] */
         h = caml_hash_mix_uint32(h, Whitehd_hd(Hd_val(v)));
         /* Copy fields into queue, not exceeding the total size [sz] */
-        for (i = 0, len = Wosize_val(v); i < len; i++) {
+        /* initialize any elements of the queue that are uninitialized */
+
+        len = Wosize_val(v);
+
+        /* we may have more fields than the maximum to use for the hash */
+        j = wr+len > sz ? sz : wr+len;
+
+        for( i = wr ; i < j; i++ ) {
+          queue[i] = Val_unit;
+        }
```

Comment:
This looks completely right, but the fact that the indexing logic is different from the loop below (wr..j instead of 0..len) meant it took me a while to read. I think it would be simpler if the two loops had the same indexing logic, e.g. by replacing the first with something like
```c
        for (i = 0; i < len; i++) {
          if (wr + i >= sz) break;
          queue[wr + i] = Val_unit;
        }
```
or equally, by replacing the second with something like
```c
        for( ; wr < j; wr++ ) {
          caml_read_field(v, i, &queue[wr]);
        }
```